### ISSUE TITLE
[Question] Is the router could be part of the broker?

### ISSUE THREAD

**Issue by MarcMagnin** [fhmq/hmq#7]
Hey,

So if I understand it, a cluster of broker require a router. That sounds like a single point of failure. Is this router functionality could be embedded in the broker such as I suppose emqtt is doing? http://emqtt.io/

Regards

**Comment by chowyu08** [comment]
I am also a bit confused about cluster design, I have another version similar to emqtt, 
branch "newbroker",  but the "router"  looks more distinct.

**Comment by MarcMagnin** [comment]
What about using raft algorithm (for instance the [etcd ](https://github.com/coreos/etcd/tree/master/raft)implementation?)

**Comment by hui6075** [comment]
it is not easy to use DHT since there's wildcard inside topic filter.. so i think simple broadcast the subscription is enough.

**Comment by MarcMagnin** [comment]
Sorry @hui6075 I'm not sure to follow. Broadcast the subscription of what to who? A broker subscription to other brokers?

**Comment by hui6075** [comment]
@MarcMagnin yes, a broker broadcast it's local clients' subscription to other brokers inside the cluster. ONLY broadcast the subscription which comes from external clients, this does work for my mosquitto cluster;)

**Comment by MarcMagnin** [comment]
They seems to have a kafka broker with raft for the clustering. That may be good to look at it at some point: https://github.com/travisjeffery/jocko

**Comment by svija** [comment]
how to save latest messages to redis? have api?
